### ISSUE TITLE
perrankseed_setup, DataBefore_inference_on_dataset

### ISSUE THREAD

**Issue by gyuilLim** [buptLwz/MABP#1]
Hi @buptLwz  

Thank you for sharing this great work.

While trying to reproduce the results from this project, I encountered several missing modules or undefined imports:

`from detectron2.engine import perrankseed_setup` : 

`from detectron2.evaluation import DataBefore_inference_on_dataset` : 

These functions dont seem to exist in the official Detectron2 repository.

If these functions are implemented in a specific version or custom fork of Detectron2, could you please let me know which version or repository you are using?

Additionally, I noticed that the tools/*process.py scripts import the following:

```python
from refer import REFER
from grefer import G_REFER
```

Could you share the source or repository for the refer and grefer modules?

Thank you in advance for your help!

**Comment by Lwzzzzzz** [comment]
Hi, thank you for your attention to our work. I'm sorry that I forgot to upload some code. However, you can refer to ReLA to obtain "refer" and "grefer". I will update the code completely later tomorrow.

**Comment by buptLwz** [comment]
Hi, thank you for your attention to our work. I'm sorry that I forgot to upload some code. However, you can refer to ReLA to obtain "refer" and "grefer". I will update the code completely later tomorrow.

**Comment by buptLwz** [comment]
Hi @gyuilLim , `Databefore_Trainer/Inference` has been removed in the new version and we have uploaded the `refer`, `grefer`, and `perrankseed_setup`  (now replaced by `default_setup`) in tools.

You can update 
`train_net.py` 
`gres_model/data_prepare.py`
`gres_model/MABP.py`
to obtain the latest version, which will not affect other parts of the model.

We also fixed some other bugs. If you have any other questions, don’t hesitate to let me know.

**Comment by gyuilLim** [comment]
Thanks for your working!

I'm just wondering about—are you using a custom version of Detectron2, particularly with modifications in `detectron2/data/build.py` and `detectron2/solver/lr_scheduler.py`?

If not, could you please let me know which version of Detectron2 you're using?

Thanks again!

**Comment by buptLwz** [comment]
@gyuilLim Thank you for your reminder. I will update the relevant instructions in the `README.md` accordingly later.

First, to minimize redundancy as much as possible, it is necessary to comment out lines 276 - 288 in `detectron2/data/build.py` when using the dataset in lmdb format. By default, `detectron2` requires `dataset_dict` to be a list of dictionaries. However, when using the lmdb format, in fact, only a list of keys needs to be passed instead of encapsulating the keys into dictionaries and then constructing a list, which will cause an error. (But for subsequent use, I will update `gres_model/data/dataset_mappers/refcoco_mapper.py` and `load_lmdb.py` in the new version so that there is no need to modify detectron2 anymore.)

Regarding `solver,` the issue may lie in `WarmupCosineRestartLR.` You can temporarily try changing `WarmupCosineRestartLR` in the config to `WarmupCosineLR,` which will not affect inference. I will update the code for `WarmupCosineRestartLR` in subsequent version.

The Detectron2 version I use is 0.6. It should be noted that I built Detectron2 from source as described in [Docs](https://detectron2.readthedocs.io/en/latest/tutorials/install.html), and there are differences from its pre-built version.

**Comment by gyuilLim** [comment]
@buptLwz 

Thanks for the clarification.  That resolves my questions!

I'll apply the temporary changes you suggested. Appreciate it.

Would it be okay to close this issue now?

**Comment by buptLwz** [comment]
Hi @gyuilLim, I have updated the code. 

You can use the new LR scheduler we implement by updating `train_net.py`, `gres_model/utils/WarmupCosineRestartLR.py`, and `configs/gres-MABP.yaml`. It theoretically has better performance than `WarmupCosineLR`, and we have provided specific introductions in `README` and comments. 

The parts related to `dataset` that were mentioned previously have also been modified, but these modifications are not necessary. We also fixed some other bugs in `train_net.py`, so you'd better update it for training.

I will close this issue. If you have any other questions, feel free to reopen it.

**Comment by gyuilLim** [comment]
Hello @buptLwz 

I’ve added a few additional details that might be helpful:

1. commented out the line :
https://github.com/buptLwz/MABP/blob/8cf89742a28f0778054631aebde7749e294194b8/gres_model/__init__.py#L15

2. set `BERT_TYPE` to `bert-base-uncased`.
https://github.com/buptLwz/MABP/blob/8cf89742a28f0778054631aebde7749e294194b8/configs/gres-MABP.yaml#L29

It might be better to avoid hard coding, if possible.

Thank you again for your great work and support!

**Comment by buptLwz** [comment]
@gyuilLim You're right. I will delete the incorrect import of `test_utils.py`. Regarding BERT,  I built BERT using a local ckpt, so the path is an absolute path. Feel free to change it to `bert-base-uncased` to directly download and import the model automatically from Hugging Face. I have also reminded users in the `README` to modify the configuration file.

**Comment by gyuilLim** [comment]
Thanks for the clarification! everything works fine now.
